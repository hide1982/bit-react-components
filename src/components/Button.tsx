import React, { forwardRef } from "react";
import styled from "styled-components";

import style from "../constants/style";
import ButtonBase from "./ButtonBase";

import type { ButtonBaseProps as Props } from "../types";

const StyledButton = styled(ButtonBase)`
  position: relative;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  height: 32px;
  padding: 0 12px;
  background-color: #aaa;
  color: #fff;
  transition: filter 300ms;

  :hover {
    filter: ${style.filter.hover};
  }

  :focus-visible {
    outline: ${style.outline};
  }

  :disabled {
    filter: ${style.filter.disabled};
    cursor: default;

    :hover {
      filter: ${style.filter.disabled};
    }
  }
`;

const Button = forwardRef<HTMLButtonElement, Props>((props, ref) => (
  <StyledButton ref={ref} {...props} />
));

export default Button;
