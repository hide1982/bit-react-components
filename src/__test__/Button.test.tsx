import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";

import Button from "../components/Button";

describe("Button Component", () => {
  test("snapshot", () => {
    render(<Button>button</Button>);
    expect(screen.getByTestId("button")).toMatchSnapshot();
  });

  test("Can use onClick event callback", () => {
    const onClick = jest.fn();
    render(<Button onClick={onClick}>hoge</Button>);

    fireEvent.click(screen.getByTestId("button"));
    expect(onClick).toBeCalledTimes(1);
  });

  test("Cannot be clicked if the button is disabled", () => {
    const onClick = jest.fn();
    render(
      <Button onClick={onClick} disabled>
        hoge
      </Button>
    );

    fireEvent.click(screen.getByTestId("button"));
    expect(onClick).toBeCalledTimes(0);
  });

  test("Can use ref", () => {
    const ref = React.createRef<HTMLButtonElement>();
    render(<Button ref={ref}>hoge</Button>);

    expect(ref.current?.dataset?.testid).toBe("button");
  });
});
