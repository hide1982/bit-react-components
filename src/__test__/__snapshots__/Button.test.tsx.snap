// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component snapshot 1`] = `
<button
  class="sc-gsTCUz hKTmWo sc-dlfnbm yryZy"
  data-testid="button"
>
  button
  <span
    class="sc-bdfBwQ eKEvNE"
    data-testid="ripple"
  />
</button>
`;
